src/edu/washington/escience/myria/operator/SamplingDistribution.java

```diff
@@ -0,0 +1,320 @@
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** The output schema. */
+  private static final Schema SCHEMA = Schema.of(ImmutableList.of(
+      Type.INT_TYPE, Type.INT_TYPE, Type.INT_TYPE, Type.STRING_TYPE),
+      ImmutableList.of("WorkerID", "StreamSize", "SampleSize", "SampleType"));
+
+  /** Total number of tuples to sample. */
+  private int sampleSize;
+
+  /** True if using a percentage instead of a specific tuple count. */
+  private boolean isPercentageSample = false;
+
+  /** Percentage of total tuples to sample. */
+  private float samplePercentage;
+
+  /** The type of sampling to perform. */
+  private final SamplingType sampleType;
+
+  /** Random generator used for creating the distribution. */
+  private Random rand;
+
+  /** Seed for the random generator. */
+  private Long randomSeed;
+
+  /**
+   * Distribution of the tuples across the workers. Value at index i == # of
+   * tuples on worker i.
+   */
+  ArrayList<Integer> tupleCounts;
+
+  /**
+   * Distribution of the actual stream size across the workers. May be different
+   * from tupleCounts if workers pre-sampled the data. Value at index i == # of
+   * tuples in stream on worker i.
+   */
+  ArrayList<Integer> streamCounts;
+
+  /** Total number of tuples across all workers. */
+  int totalTupleCount = 0;
+
+  private SamplingDistribution(Operator child, SamplingType sampleType,
+      Long randomSeed) {
+    super(child);
+    this.sampleType = sampleType;
+    this.randomSeed = randomSeed;
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a specific sample size.
+   *
+   * @param sampleSize
+   *          total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, int sampleSize,
+      SamplingType sampleType, Long randomSeed) {
+    this(child, sampleType, randomSeed);
+    this.sampleSize = sampleSize;
+    Preconditions.checkState(this.sampleSize >= 0,
+        "Sample Size must be >= 0: %s", this.sampleSize);
+  }
+
+  /**
+   * Instantiate a SamplingDistribution operator using a percentage of total
+   * tuples.
+   *
+   * @param samplePercentage
+   *          percentage of total samples to create a distribution for.
+   * @param sampleType
+   *          the type of sampling distribution to create
+   * @param child
+   *          extracts (WorkerID, PartitionSize, StreamSize) information from
+   *          this child.
+   * @param randomSeed
+   *          value to seed the random generator with. null if no specified seed
+   */
+  public SamplingDistribution(Operator child, float samplePercentage,
```

Comment:
I see your point here. Part of the reasoning here is to make it easier for RACO to know the difference between specifying tupleSize vs percentSize. I think I'll leave it as-is for the time-being but a good discussion to have at some point. (it'll be straightforward to modify the code)